File: backend/api/schemas/violation.py
from typing import Optional, List
from uuid import UUID
from datetime import datetime
from pydantic import BaseModel, Field, PositiveFloat


class ViolationBase(BaseModel):
    record_id: UUID
    timestamp: datetime
    vehicle_type: Optional[str]
    violation_type: str
    license_plate: Optional[str]
    confidence: Optional[float]
    frame_image_path: str
    vehicle_image_path: str
    lp_image_path: Optional[str]
    status: Optional[str] = "pending"
    
    model_config = {
        "from_attributes": True
    }

class ViolationCreate(ViolationBase):
    pass

class ViolationUpdate(BaseModel):
    vehicle_type: Optional[str] = None
    violation_type: Optional[str] = None
    license_plate: Optional[str] = None
    confidence: Optional[float] = None
    status: Optional[str] = None

    model_config = {
        "from_attributes": True
    }

class ViolationOut(ViolationBase):
    id: UUID
    version_id: Optional[UUID]

    model_config = {
        "from_attributes": True
    }

class ViolationVersionOut(BaseModel):
    id: UUID
    violation_id: UUID
    officer_id: UUID
    updated_at: datetime
    change_type: str
    details: Optional[dict]

    model_config = {
        "from_attributes": True
    }

class Pagination(BaseModel):
    page: int
    limit: int
    total: int
    totalPages: int

    model_config = {
        "from_attributes": True
    }

class PaginatedViolations(BaseModel):
    pagination: Pagination
    data: List[ViolationOut]

    model_config = {
        "from_attributes": True
    }

File: frontend/lib/types/api/v1/violation.ts
import { Camera } from "./camera"
import { Officer } from "./officer"
import { Pagination } from "./pagination"

export interface Violation {
    id: string
    camera_id: string
    timestamp: string
    vehicle_type: string | null
    violation_type: string
    license_plate: string | null
    confidence: number | null
    status: string | null
    frame_image_path: string
    vehicle_image_path: string
    lp_image_path: string | null
    version_id: string | null
    camera?: Camera
}

export interface PaginatedViolations {
    pagination: Pagination
    data: Violation[]
}

export interface ViolationHistory {
    id: string
    violation_id: string
    officer_id: string
    timestamp: string
    vehicle_type: string | null
    violation_type: string
    license_plate: string | null
    confidence: number | null
    frame_image_path: string
    vehicle_image_path: string
    lp_image_path: string | null
    updated_at: string
    change_type: string
    notes?: string | null     
    source_id?: string | null  
    camera?: Camera
    officer?: Officer
    status: string | null
}
  

export interface ViolationUpdate {
    vehicle_type: string | null
    violation_type: string | null
    license_plate: string | null
    confidence: number | null
    status: string | null
    notes?: string | null 
}

export interface GetViolationsParams {
    page?: number
    limit?: number
    status?: string | null
    violation_type?: string | null
    search?: string | null
    camera_id?: string | null
    vehicle_type?: string | null
    confidence_min?: number | null
    confidence_max?: number | null
    timestamp_from?: string | null
    timestamp_to?: string | null
  }